components/tel/Core/Src/freertos.c

```diff
@@ -0,0 +1,462 @@
+    // enter command mode
+    uint8_t tripleplus[3] = {'+', '+', '+'};
+
+    osDelay(1000);
+    HAL_UART_Transmit(&huart3, tripleplus, sizeof(tripleplus), 1000);
+    osDelay(2000);
+    
+    uint8_t check_dl[5] = "ATDL\r";
+    HAL_UART_Transmit(&huart3, check_dl, sizeof(check_dl), 1000);
+    
+    osDelay(1000);
+
+    // send destination low change AT command
+    uint8_t change_dl[15] = "ATDL 418C9CBD\r";
+    HAL_UART_Transmit(&huart3, change_dl, sizeof(change_dl), 1000);
+
+    osDelay(1000);
+
+    // send ATDL
+    HAL_UART_Transmit(&huart3, check_dl, sizeof(check_dl), 1000);
+    
+    osDelay(1000);
+    
+    // send "ATAC" (apply changes)
+    uint8_t apply_changes[5] = "ATAC\r";
+    HAL_UART_Transmit(&huart3, apply_changes, sizeof(apply_changes), 1000);
+
+    osDelay(1000);
+
+    // send "ATWR" (write changes to non-volatile memory)
+    uint8_t write_changes[5] = "ATWR\r";
+    HAL_UART_Transmit(&huart3, write_changes, sizeof(write_changes), 1000);
+
+    osDelay(1000);
+
+    HAL_GPIO_WritePin(KERNEL_LED_GPIO_Port, KERNEL_LED_Pin, GPIO_PIN_SET);
+
+    osThreadExit();
+  }
+}
+
+void sendChar(char c)
+{
+  HAL_StatusTypeDef uart_tx_status;
+
+  //Convert to ASCII
+  uint8_t c_H = "0123456789ABCDEF"[(c >> 4) & 0xFUL];
+  uint8_t c_L = "0123456789ABCDEF"[ c & 0xFUL];
+
+  uart_tx_status = HAL_UART_Transmit(&huart3, &c_H, 1, 1000);
+  uart_tx_status = HAL_UART_Transmit(&huart3, &c_L, 1, 1000);
+}
+
+__NO_RETURN void StartReadIMU(void *argument){
+  union FloatBytes gy_x, gy_y, gy_z, ax_x, ax_y, ax_z;
+  /* Infinite loop */
+  while(1)
+  {
+    gy_x.float_value = gyro(GYRO_X);
+    gy_y.float_value = gyro(GYRO_Y);
+    gy_z.float_value = gyro(GYRO_Z);
+    ax_x.float_value = accel(ACCEL_X);
+    ax_y.float_value = accel(ACCEL_Y);
+    ax_z.float_value = accel(ACCEL_Z);
+
+    // IMU DATA
+    addtoIMUQueue("G", "X", gy_x);
+    addtoIMUQueue("G", "Y", gy_y);
+    addtoIMUQueue("G", "Z", gy_z);
+    addtoIMUQueue("A", "X", ax_x);
+    addtoIMUQueue("A", "Y", ax_y);
+    addtoIMUQueue("A", "Z", ax_z);
+
+    osDelay(1000);
+  }
+
+  // In case we accidentally exit from task loop
+  osThreadTerminate(NULL);
+
+}
+
+void addtoIMUQueue(char* type, char* dimension, union FloatBytes data){
+    IMU_msg_t imu_message;
+
+    imu_message.imu_type = type[0];
+    imu_message.dimension = dimension[0];
+    for (int i = 0; i < 4; i++) {
+        imu_message.data[i] = data.bytes[i];
+    }
+
+    osMessageQueuePut(imuMessageQueueHandle, &imu_message, 0U, 0U);
+}
+
+__NO_RETURN void StartTransmitIMU(void *argument){
+  osStatus_t imu_queue_status;
+  IMU_msg_t imu_message;
+
+  while(1)
+  {
+    // Check if there are messages in the queue
+    if (osMessageQueueGetCount(imuMessageQueueHandle) == 0) {
+        osThreadYield();
+    }
+
+    imu_queue_status = osMessageQueueGet(imuMessageQueueHandle, &imu_message, NULL, osWaitForever);
+
+    if (imu_queue_status != osOK){
+      osThreadYield();
+    }
+    uint8_t imu_buffer[9] = {0};
```

Comment:
Same as CAN buffer, 9 would be better as a macro. It's a magic number without an initially apparent reason as to why it is size 9.